docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+
+```
+IDemoGetType    | IDemoStoreType
+==================================
+QueryInterface  | QueryInterface
+AddRef          | AddRef
+Release         | Release
+GetString       | StoreString
+```
+
+Looking at `DemoImpl`, we already have an implementation for `GetString()` and `StoreString()`, but what about the `IUnknown` functions? How to [implement an `IUnknown` instance][doc_impliunknown] is beyond the scope of this tutorial, but it can be done manually in `ComWrappers`. However, in this tutorial, you'll let the runtime handle that part. You can get the `IUnknown` implementation using the [`ComWrappers.GetIUnknownImpl()`][api_comwrappers] method.
+
+It might seem like you've implemented all the methods, but unfortunately, only the `IUnknown` functions are consumable in a COM vtable. Since COM is outside of the runtime, you'll need to create native function pointers to your `DemoImpl` implementation. This can be done easily using C# function pointers and the [`UnmanagedCallersOnlyAttribute`][api_unmanagedcallersonly]. You can create a function to insert into the vtable by creating a `static` function that mimics the COM function signature. Following is an example of the COM signature for `IDemoGetType.GetString()` &ndash; recall from the COM ABI that the first argument is the instance itself.
+
+```csharp
+[UnmanagedCallersOnly]
+public static int GetString(IntPtr _this, IntPtr* str);
+```
+
+The wrapper implementation of `IDemoGetType.GetString()` should consist of marshalling logic and then a dispatch to the managed object being wrapped. All the state for dispatch is contained within the provided `_this` argument. The `_this` argument will actually be of type `ComInterfaceDispatch*`. This type represents a low-level structure with a single field, `Vtable`, that will be discussed later. Further details of this type and its layout are an implementation detail of the runtime and should not be depended upon. In order to retrieve the managed instance from a `ComInterfaceDispatch*` instance, use the following code:
+
+```csharp
+IDemoGetType inst = ComInterfaceDispatch.GetInstance<IDemoGetType>((ComInterfaceDispatch*)_this);
+```
+
+Now that you have a C# method that can be inserted into a vtable, you can construct the vtable. Note the use of [`RuntimeHelpers.AllocateTypeAssociatedMemory()`][api_allocatetypeassociatedmemory] for allocating memory in a way that works with [unloadable][doc_unloadability] assemblies.
+
+```csharp
+GetIUnknownImpl(
+    out IntPtr fpQueryInterface,
+    out IntPtr fpAddRef,
+    out IntPtr fpRelease);
+
+{
+    int tableCount = 4;
+    int idx = 0;
+    var vtable = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(
+        typeof(DemoComWrappers),
+        IntPtr.Size * tableCount);
+    vtable[idx++] = fpQueryInterface;
+    vtable[idx++] = fpAddRef;
+    vtable[idx++] = fpRelease;
+    vtable[idx++] = (IntPtr)(delegate* unmanaged<IntPtr, IntPtr*, int>)&ABI.
```

Comment:
This is more of a C# rant - why is it necessary to type out the "native delegate" type here? The funny thing is that the compiler knows exactly what it should be, since if I write it wrong it will complain. It would be nice if just `(IntPtr)&Method` worked.

Reply:
We should take this offline. There is actually a reason and it relates to our calling convention extension model.